src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+        };
+
+        match e {
+            // See `connect` above for the rationale behind `forget`
+            Ok(_) => {
+                io.read = State::Pending(buf, 0); // 0 is ignored on read side
+                mem::forget(me.clone());
+                true
+            }
+
+            // If ERROR_PIPE_LISTENING happens then it's not a real read error,
+            // we just need to wait for a connect.
+            Err(ref e) if e.raw_os_error() == Some(ERROR_PIPE_LISTENING as i32) => false,
+
+            // If some other error happened, though, we're now readable to give
+            // out the error.
+            Err(e) => {
+                io.read = State::Err(e);
+                if let Some(waker) = io.read_waker.take() {
+                    waker.wake();
+                }
+                true
+            }
+        }
+    }
+
+    fn schedule_write(me: &Arc<Inner>, buf: Vec<u8>, pos: usize, io: &mut Io) {
+        // Very similar to `schedule_read` above, just done for the write half.
+        let e = unsafe {
+            let overlapped = me.write.as_mut_ptr() as *mut _;
+            me.handle.write_overlapped(&buf[pos..], overlapped)
+        };
+
+        match e {
+            // See `connect` above for the rationale behind `forget`
+            Ok(_) => {
+                io.write = State::Pending(buf, pos);
+                mem::forget(me.clone())
+            }
+            Err(e) => {
+                io.write = State::Err(e);
+                if let Some(waker) = io.write_waker.take() {
+                    waker.wake();
+                }
+            }
+        }
+    }
+
+    fn post_register(me: &Arc<Inner>) {
+        let mut io = me.io.lock().unwrap();
+        if Inner::schedule_read(&me, &mut io) {
+            if let State::None = io.write {
+                if let Some(waker) = io.write_waker.take() {
+                    waker.wake();
+                }
+            }
+        }
+    }
+
+    fn get_buffer(&self) -> Vec<u8> {
+        self.pool.lock().unwrap().get(8 * 1024)
+    }
+
+    fn put_buffer(&self, buf: Vec<u8>) {
+        self.pool.lock().unwrap().put(buf)
+    }
+}
+
+unsafe fn cancel<T: AsRawHandle>(handle: &T, overlapped: &Overlapped) -> io::Result<()> {
+    let ret = CancelIoEx(handle.as_raw_handle(), overlapped.as_mut_ptr() as *mut _);
+    if ret == 0 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
+
+fn connect_done(status: &OVERLAPPED_ENTRY) {
+    let status = CompletionStatus::from_entry(status);
+
+    // Acquire the `Arc<Inner>`. Note that we should be guaranteed that
+    // the refcount is available to us due to the `mem::forget` in
+    // `connect` above.
+    let me = unsafe { overlapped2arc!(status.overlapped(), Inner, connect) };
+
+    // Flag ourselves as no longer using the `connect` overlapped instances.
+    let prev = me.connecting.swap(false, SeqCst);
+    assert!(prev, "wasn't previously connecting");
```

Comment:
Can we include `mio::windows::NamedPipe` in the message if it does ever come up?